block/gs/adapter.go

```diff
@@ -132,27 +132,31 @@ func (a *Adapter) Get(obj block.ObjectPointer, _ int64) (io.ReadCloser, error) {
 	return r, err
 }
 
-func (a *Adapter) List(lsOpt block.ListOpts) ([]string, error) {
+func (a *Adapter) Walk(walkOpt block.WalkOpts, walkFn block.WalkFunc) error {
 	var err error
 	defer reportMetrics("List", time.Now(), nil, &err)
-	qualifiedPrefix, err := resolveNamespacePrefix(lsOpt)
+	qualifiedPrefix, err := resolveNamespacePrefix(walkOpt)
 	if err != nil {
-		return nil, err
+		return err
 	}
 
-	var keys []string
 	iter := a.client.Bucket(qualifiedPrefix.StorageNamespace).Objects(context.Background(), &storage.Query{Prefix: qualifiedPrefix.Prefix})
+
 	for {
 		attrs, err := iter.Next()
+
 		if errors.Is(err, iterator.Done) {
 			break
 		}
 		if err != nil {
-			return nil, fmt.Errorf("bucket(%s).Objects(): %w", qualifiedPrefix.StorageNamespace, err)
+			return fmt.Errorf("bucket(%s).Objects(): %w", qualifiedPrefix.StorageNamespace, err)
+		}
+
+		if err := walkFn(attrs.Name); err != nil {
+			return fmt.Errorf("walkFn failed: %w", err)
```

Comment:
can we return the original error in this case? it will enable the user of the able to stop the processing by returning EOF - and I don't think we need to wrap in this case.

Reply:
(relevant to all adapters)

Reply:
Done.